File: web/frontend/src/api/letterboxd.ts
import { apiJson } from "./client";
import type { AppConfigResponse, AppConfigUpdate, AuthCheck, DeepAuthCheck } from "./types";

export async function checkAuth(): Promise<AuthCheck> {
  return apiJson("/api/auth/check");
}

export async function deepCheckAuth(): Promise<DeepAuthCheck> {
  return apiJson("/api/auth/deep-check");
}

export async function updateAppConfig(config: AppConfigUpdate): Promise<AppConfigResponse> {
  return apiJson("/api/auth/config", {
    method: "POST",
    body: JSON.stringify(config),
  });
}

export async function addToWatchlist(slug: string, remove = false) {
  return apiJson("/api/letterboxd/watchlist", {
    method: "POST",
    body: JSON.stringify({ slug, remove, confirmed: true }),
  });
}

export async function markWatched(slug: string, remove = false) {
  return apiJson("/api/letterboxd/watched", {
    method: "POST",
    body: JSON.stringify({ slug, remove, confirmed: true }),
  });
}

export async function toggleLike(slug: string, remove = false) {
  return apiJson("/api/letterboxd/like", {
    method: "POST",
    body: JSON.stringify({ slug, remove, confirmed: true }),
  });
}

export async function rateFilm(slug: string, rating: number) {
  return apiJson("/api/letterboxd/rate", {
    method: "POST",
    body: JSON.stringify({ slug, rating, confirmed: true }),
  });
}

export async function writeReview(slug: string, review: string, rating?: number) {
  return apiJson("/api/letterboxd/review", {
    method: "POST",
    body: JSON.stringify({ slug, review, rating, confirmed: true }),
  });
}

File: web/backend/app/api/letterboxd.py
from __future__ import annotations

from fastapi import APIRouter

from ..schemas.letterboxd import (
    ConfirmedSlugRequest,
    LetterboxdActionResponse,
    RateRequest,
    ReviewRequest,
)
from ..services.letterboxd_service import call_write_tool


router = APIRouter(prefix="/api/letterboxd", tags=["letterboxd"])


@router.post("/watchlist", response_model=LetterboxdActionResponse)
async def add_to_watchlist(request: ConfirmedSlugRequest) -> LetterboxdActionResponse:
    result = await call_write_tool(
        "add_to_watchlist",
        {"slug": request.slug, "remove": request.remove},
        request.confirmed,
    )
    return LetterboxdActionResponse(ok=True, tool="add_to_watchlist", result=result)


@router.post("/watched", response_model=LetterboxdActionResponse)
async def add_to_watched(request: ConfirmedSlugRequest) -> LetterboxdActionResponse:
    result = await call_write_tool(
        "add_to_watched",
        {"slug": request.slug, "remove": request.remove},
        request.confirmed,
    )
    return LetterboxdActionResponse(ok=True, tool="add_to_watched", result=result)


@router.post("/like", response_model=LetterboxdActionResponse)
async def toggle_like(request: ConfirmedSlugRequest) -> LetterboxdActionResponse:
    result = await call_write_tool(
        "toggle_like",
        {"slug": request.slug, "remove": request.remove},
        request.confirmed,
    )
    return LetterboxdActionResponse(ok=True, tool="toggle_like", result=result)


@router.post("/rate", response_model=LetterboxdActionResponse)
async def rate_film(request: RateRequest) -> LetterboxdActionResponse:
    result = await call_write_tool(
        "rate_film",
        {"slug": request.slug, "rating": request.rating},
        request.confirmed,
    )
    return LetterboxdActionResponse(ok=True, tool="rate_film", result=result)


@router.post("/review", response_model=LetterboxdActionResponse)
async def write_review(request: ReviewRequest) -> LetterboxdActionResponse:
    args: dict = {"slug": request.slug, "review": request.review}
    if request.rating:
        args["rating"] = request.rating
    result = await call_write_tool("write_review", args, request.confirmed)
    return LetterboxdActionResponse(ok=True, tool="write_review", result=result)
